### ISSUE TITLE
Add SceneTree signals for adding/removing a Node to/from a group.

### ISSUE THREAD

**Issue by willnationsdev** [godotengine/godot-proposals#1259]
**Describe the project you are working on:**
An RPG.

**Describe the problem or limitation you are having in your project:**
I use nodes and groups as a sort of state machine for system-wide process callbacks, very similar to how the Viewport node works to execute input callbacks, only, instead of each node implementing an input callback, a System node executes an operation over every node in the collection.

This works great for iterative operations like process, but not great for in-place state-change operations, e.g. going from state A to state B should trigger some logic (represented as removing group A and adding group B). I could code this logic in the individual node's script, but then I'd be sharing logic across several scripts.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**

What I would prefer is if I could have a single callback that handles every node that is undergoing a group state change. Since groups are actually maintained by the SceneTree anyway, it makes sense that the SceneTree should have a signal for handling an update to its `group_map` data. Then, I could just connect to that signal and handle every node's mutation logic in a single place.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**

1. Add `group_changed` signal to SceneTree with a reference to the node that was added or removed to/from the group and a string for which group was updated.
1. Emit the signal in `SceneTree::add_to_group()` and `SceneTree::remove_from_group()`.
1. Add signal to SceneTree's `_bind_methods()` function.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

It can be worked around, but not in an architecturally clean way. At least not without traits (though that wouldn't be optimal either since it would only work with GDScript and not other languages).

**Is there a reason why this should be core and not an add-on in the asset library?:**
Can't be fixed with a plugin or asset.

**Comment by TheDuriel** [comment]
I don't quite agree with why you want this. But w/e.

Care must be taken that the signal is only emitted when the User changes groups. As hidden groups are heavily used within the editor, and some specialized nodes. And you could quickly be spammed by hundreds if not thousands of calls.

**Comment by Xrayez** [comment]
> It can be worked around, but not in an architecturally clean way. At least not without traits (though that wouldn't be optimal either since it would only work with GDScript and not other languages).

Interestingly, I use "component" nodes which describe the "entity" traits (root node). And those components can be added to group(s). Retrieving all objects who share the same trait is a matter of `get_nodes_in_group` call, and then you just fetch the `owner` to get the root node. One could in theory remove and add those traits on demand, and this is where signals could be used then, if I understand correctly.

**Comment by willnationsdev** [comment]
@Xrayez 

> One could in theory remove and add those traits on demand, and this is where signals could be used then, if I understand correctly.

Yes, you could, in theory, use the adding and removing of groups to trigger the adding and removing of component nodes. Though, I wouldn't necessarily recommend that since that's pretty magical and there wouldn't be a way to pass initialization logic alongside the state change. I'm just wanting to trigger function executions for `on_add` and `on_remove` that have to do with the system itself and not the actual components (like updating db indexes of often-used component combinations for optimization purposes, etc.).

**Comment by willnationsdev** [comment]
@TheDuriel 

Perhaps this would have to wait until we get an project-wide groups configuration system up and running? That way the callback would only execute for the user-defined groups configured there? Either that, or switch to a naming convention thing where groups with an underscore prefix are not added to the callbacks (that sounds much simpler and easier to work with).

**Comment by me2beats** [comment]
I have a case when I would like to know that a node of some group is created or added to the tree, but I don't want to set a script on the node.
So `node_added_to_group(group:String,node:Node) ` signal would be useful.

**Comment by aXu-AP** [comment]
In addition to having this in SceneTree, I think Node should also have these signals: `group_entered(group)`, `group_exited(group)`.

**Comment by unfallible** [comment]
I think this is absolutely essential functionality for groups, as the engine currently makes it very difficult to monitor the nodes within a given group.

However, I strongly feel that rather than giving the tree signals for whenever a node is added to one of its groups (e.g. `signal node_added_to_group(node: Node, group_name: String)` and `signal node_removed_from_group(node: Node, group_name: String)`), the SceneTree should have signals for each group, which would be managed through register and unregister functions on the tree itself (e.g. `func connect_node_added_to_group(group: String, callback: Callable) -> void`, `func disconnect_node_added_to_group(group: String, callback: Callable) -> void`, etc).
Outside of some very specific use cases such as plugins, almost all nodes that would want to listen for changes in group membership are only going to be interested in particular groups. Consequently, I believe that attaching the signals to the groups themselves has two main advantages.
First, it would be better for performance. If the signals are shared by every group, then the callbacks for every node that is monitoring a group will be triggered every single time any node's group membership changes. This would be potentially catastrophic for performance.
Second, I believe there are significant ergonomic advantages to tying the signals to particular groups. Consequently, most of the actual callbacks are going to begin with some line of code such as
```
func on_node_added_to_group(node: Node, group: String) -> void:
	if group != "relevant_group":
		return
	[...]
```
That's uglier, requires more typing, *and* it requires hard coding the name of the relevant group, which will make code harder to maintain.

**Comment by Zireael07** [comment]
> Outside of some very specific use cases such as plugins, almost all nodes that would want to listen for changes in group membership are only going to be interested in particular groups. Consequently, I believe that attaching the signals to the groups themselves has two main advantages.

I agree completely
